std/mutex.zig

```diff
@@ -5,74 +5,145 @@ const AtomicRmwOp = builtin.AtomicRmwOp;
                 return Held{ .mutex = self };
             if (c != 2)
-                c = @atomicRmw(i32, &self.linux_lock, AtomicRmwOp.Xchg, 2, AtomicOrder.Acquire);
+                c = @atomicRmw(i32, &self.lock, AtomicRmwOp.Xchg, 2, AtomicOrder.Acquire);
             while (c != 0) {
-                const rc = linux.futex_wait(&self.linux_lock, linux.FUTEX_WAIT | linux.FUTEX_PRIVATE_FLAG, 2, null);
+                const rc = linux.futex_wait(&self.lock, linux.FUTEX_WAIT | linux.FUTEX_PRIVATE_FLAG, 2, null);
                 switch (linux.getErrno(rc)) {
                     0, linux.EINTR, linux.EAGAIN => {},
                     linux.EINVAL => unreachable,
                     else => unreachable,
                 }
-                c = @atomicRmw(i32, &self.linux_lock, AtomicRmwOp.Xchg, 2, AtomicOrder.Acquire);
+                c = @atomicRmw(i32, &self.lock, AtomicRmwOp.Xchg, 2, AtomicOrder.Acquire);
             }
-        } else {
-            _ = self.spin_lock.acquire();
+            return Held { .mutex = self };
         }
-        return Held{ .mutex = self };
-    }
+    },
+    builtin.Os.windows => struct {
+
+        lock: ?windows.CRITICAL_SECTION,
+        init_once: windows.PINIT_ONCE,
+
+        pub const Held = struct {
+            mutex: *Mutex,
+
+            pub fn release(self: Held) void {
+                if (self.mutex.lock) |*lock| {
+                    windows.LeaveCriticalSection(lock);
+                }
+            }
+        };
+
+        pub fn init() Mutex {
+            return Mutex { 
+                .lock = null,
+                .init_once = undefined,
+            };
+        }
+
+        extern fn initCriticalSection(InitOnce: *windows.PINIT_ONCE, Parameter: ?windows.PVOID, Context: ?windows.PVOID) windows.BOOL {
+            if (Context) |ctx| {
+                var mutex = @ptrCast(*?windows.CRITICAL_SECTION, @alignCast(8, ctx));
+                if (mutex.* == null) {
+                    var lock: windows.CRITICAL_SECTION = undefined;
+                    windows.InitializeCriticalSection(&lock);
+                    mutex.* = lock;
+                }
+                return windows.TRUE;
+            }
+            return windows.FALSE;
+        }
+
+        pub fn deinit(self: *Mutex) void {
+            if (self.lock) |*lock| {
+                windows.DeleteCriticalSection(lock);
+            }
+        }
+
+        pub fn acquire(self: *Mutex) Held {
+            if (self.lock) |*lock| {
+                windows.EnterCriticalSection(lock);
+            } else {
+                var mutex = @ptrCast(?windows.PVOID, self);
+                if (windows.InitOnceExecuteOnce(&self.init_once, initCriticalSection, null, mutex) == windows.TRUE) {
+                    windows.EnterCriticalSection(&self.lock.?);
+                } else {
+                    @panic("unable to initialize Mutex");
```

Comment:
It looks like the possible error that InitOnceExecuteOnce returns is only from your callback function. So your callback function should never return an error, and instead of `@panic` here it should be `unreachable` (or assert that the return value of InitOnceExecuteOnce is nonzero.

Reply:
so could `initCriticalSection` look like this?
```
extern fn initCriticalSection(InitOnce: *windows.RTL_RUN_ONCE, Parameter: ?windows.PVOID, Context: ?windows.PVOID) windows.BOOL {
    var lock = @ptrCast(*windows.CRITICAL_SECTION, @alignCast(@alignOf(*windows.CRITICAL_SECTION), ctx.?));
    windows.InitializeCriticalSection(lock);
    return windows.TRUE;
}
```